File: CS/XpoXtraReports/XpoBaseReport.cs
using System;
using System.Drawing;
using System.Collections;
using System.ComponentModel;
using DevExpress.XtraReports.UI;
using DevExpress.Xpo;

namespace XpoXtraReports
{
    public partial class XpoBaseReport : DevExpress.XtraReports.UI.XtraReport
    {
        public XpoBaseReport()
        {
            InitializeComponent();
        }

        public XpoBaseReport(Session session)
            : this() {
            _XpoSession = session;
        }

        private void CreateDataSource() {
            if(XpoSession != null) {
                Type xpoType = Type.GetType(XpoClassName, true);
                this.DataSource = new XPCollection(XpoSession, XpoSession.GetClassInfo(xpoType));
            }
        }

        private Session _XpoSession;
        [Browsable(false)]
        public Session XpoSession {
            get {
                return _XpoSession;
            }
            set {
                this._XpoSession = value;
                this.CreateDataSource();
            }
        }

        [Browsable(false)]
        public XPBaseCollection Collection
        {
            get
            {
                return this.DataSource as XPBaseCollection;
            }
        }

        private string _XpoClassName;
        public string XpoClassName {
            get { return _XpoClassName; }
            set {
                if(this._XpoClassName != value) {
                    this._XpoClassName = value;
                    this.CreateDataSource();
                }
            }
        }        

        protected override void SerializeProperties(DevExpress.XtraReports.Serialization.XRSerializer serializer) {
            base.SerializeProperties(serializer);
            serializer.SerializeString("XpoClassName", this.XpoClassName);
        }
        protected override void DeserializeProperties(DevExpress.XtraReports.Serialization.XRSerializer serializer) {
            base.DeserializeProperties(serializer);
            this.XpoClassName = serializer.DeserializeString("XpoClassName", "");
        }
    }
}

File: VB/XpoXtraReports/XpoBaseReport.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Drawing
Imports System.Collections
Imports System.ComponentModel
Imports DevExpress.XtraReports.UI
Imports DevExpress.Xpo

Namespace XpoXtraReports
	Partial Public Class XpoBaseReport
		Inherits DevExpress.XtraReports.UI.XtraReport
		Public Sub New()
			InitializeComponent()
		End Sub

		Public Sub New(ByVal session As Session)
			Me.New()
			_XpoSession = session
		End Sub

		Private Sub CreateDataSource()
			If XpoSession IsNot Nothing Then
				Dim xpoType As Type = Type.GetType(XpoClassName, True)
				Me.DataSource = New XPCollection(XpoSession, XpoSession.GetClassInfo(xpoType))
			End If
		End Sub

		Private _XpoSession As Session
		<Browsable(False)> _
		Public Property XpoSession() As Session
			Get
				Return _XpoSession
			End Get
			Set(ByVal value As Session)
				Me._XpoSession = value
				Me.CreateDataSource()
			End Set
		End Property

		<Browsable(False)> _
		Public ReadOnly Property Collection() As XPBaseCollection
			Get
				Return TryCast(Me.DataSource, XPBaseCollection)
			End Get
		End Property

		Private _XpoClassName As String
		Public Property XpoClassName() As String
			Get
				Return _XpoClassName
			End Get
			Set(ByVal value As String)
				If Me._XpoClassName <> value Then
					Me._XpoClassName = value
					Me.CreateDataSource()
				End If
			End Set
		End Property

		Protected Overrides Sub SerializeProperties(ByVal serializer As DevExpress.XtraReports.Serialization.XRSerializer)
			MyBase.SerializeProperties(serializer)
			serializer.SerializeString("XpoClassName", Me.XpoClassName)
		End Sub
		Protected Overrides Sub DeserializeProperties(ByVal serializer As DevExpress.XtraReports.Serialization.XRSerializer)
			MyBase.DeserializeProperties(serializer)
			Me.XpoClassName = serializer.DeserializeString("XpoClassName", "")
		End Sub
	End Class
End Namespace
